### ISSUE TITLE
Blockified Single Product Template: Create the Upsell Products block

### ISSUE THREAD

**Issue by gigitux** [woothemes/woocommerce#42358]
@xristos3490 noticed that upsell products are not visible on the blockified Single Products (even if the product has those).

## To reproduce

1. Edit a product and add upsell products.
2. Ensure that you are using the classic Single Product Template. 
3. Visit the product on the frontend.
4. Verify that the upsell products are visible.
5. Enable the blockified Single Product Template.
6. Visit the product on the frontend.
7. The upsell products are not visible.

|Classic Single Product Template| Blockified Single Product Template |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/3718142e-bdfc-4a42-a09a-158410ce16cb)|![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/b8deda49-bac7-4763-83a8-2bd67bda9817)|

The reason is that we didn't create a block to show the upsell Products. We should create a block similar to the `Related Products` block, but for the Upsells Products.

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/98fa20d7-84d1-498f-ba43-b73a451d8b0c)

The Upsells Products will be composed of a main wrapper called `Upsell Products` and a child block: `Upsell Products block` (like the Related Products block). Furthermore, on the sidebar, we should display a notice that says that the products will depend on the product configuration, and the block may display nothing on the front end if the product has no upsell products.

@shaunandrews, what do you think?

**Comment by manospsyx** [comment]
Hey folks,

Let's tread carefully here! This problem is a great opportunity to set an example for all Extensions out there to follow.

- "Up-sells" are defined at a product level: Some products may have up-sells defined by the merchant, and some not.
- This feature is very similar to many features you'll find in Extensions: It's common to have a field or form used to add more functionality or options in the Single Product and other pages: What the shopper will see is not going to be consistently there for all products, but it might be on those products that utilize the feature. 

The challenge here is that:

- On the one hand, the merchant may want to visually customize these features for all products on their store -- e.g. add margins around gift card fields/forms or product add-on fields (which should appear in a specific location above the add-to-cart button) or move up-sells up or down on the page. 
- On the other hand, it does not make sense for all these features to appear and be customizable in the Single Product Editor template, because (i) dozens of Extensions adding elements to product pages might be installed; and (ii) not every product will utilize all the features. 

In this specific case, it seems odd to me that Up-Sells (added manually by merchants) are lumped together into the same block as Related Products (added by Woo Core to every product automatically, can be turned via a filter only). 

As a merchant, I would like to 

- know that I can add up-sells to my products, and 
- move them before or after Related Products. 

And therefore, it initially seems to make sense to have them in the Site Editor. However, I am saying this knowing that:

- it's unsustainable to have every product feature represented by a Block in the Single Product template;
- some features may not make any sense in the same template -- for a example a product can't be both a gift card and have bundled items.

We eventually need to decide how much it makes sense to let merchants customize in the Single Product template, and if we limit the template to contain only things that are shared among every product (and product type) out there, then figure out what will be the best way (and place) for merchants to customize the appearance of elements like up-sells, add-ons, gift cards and bundled products.

**Comment by gigitux** [comment]
Hey @manospsyx! Thank you for your great observations! 

> In this specific case, it seems odd to me that Up-Sells (added manually by merchants) are lumped together into the same block as Related Products (added by Woo Core to every product automatically, can be turned via a filter only).

This is a good point.

> We eventually need to decide how much it makes sense to let merchants customize in the Single Product template, and if we limit the template to contain only things that are shared among every product (and product type) out there, then figure out what will be the best way (and place) for merchants to customize the appearance of elements like up-sells, add-ons, gift cards and bundled products.

I understand your points and I agree. Should the product editor take care of this kind of customization? Let me tag folks that should be involved in this discussion @tjcafferkey @danielwrobert @nerrad @shaunandrews 

Furthermore, I created a PR to fix this issue (at least for now) without using a block, but restoring the action `woocommerce_upsell_display`.

**Comment by manospsyx** [comment]
> Should the product editor take care of this kind of customization?

I'm not sure if there's an obvious answer here :) 

Re Up-Sells / Related Products:

- We could show both blocks (Related, Up-sells) and let the user move them around and customize them at will, but if we do so, it should be clear to the user that (i) some product pages may not have up-sells to render in the Up-Sells block at all, and (ii) the actual contents of both blocks will vary depending on the product. 
- Do we want to give merchants the option to hide Related Products globally here? Should that be done by removing the entire block? 

Re the boundaries of the Site Editor:

I am tempted to think that it would be useful to introduce product-type templates. 

Product-type templates could allow users to customize variable products, or bundles, or composite products, and:

- give the user access to more blocks relevant to the specific product type;
- let the user see a more representative version of the add-to-cart block, richer in inner blocks.

But should developers also be able to expose "conditional templates" to the user? For example 

- a "Single Product template for products with add-ons"

If developers are able to expose entire conditional templates at will, what should happen when multiple conditions are satisfied at the same time? Which template takes precedence if a product is both a variable one, and offers add-ons? 

Could template parts perhaps serve the need to customize Extension-injected content globally?

These questions are, of course, rhetorical, but they perhaps suggest that there might not be a one-size fits-all solution. WooCommerce may treat the add-to-cart block as a "black box" in the Single Product template, but at the same time always show an Up-Sells block and let the user know that it will only render something if up-sells are defined at the product level.

**Comment by nerrad** [comment]
Manos has raised very valid points. A few thoughts:
- @jarekmorawski I think this conversation is relevant to explorations you've been considering in the editor (I think I saw one around adding external links to CTA buttons) and might be a great use-case for things like the "upsell" configuration in the editor where the design of the upsell block could potentially be modified.
- Beyond customizable design of the feature itself, the other value for some features (such as the upsell, related products etc) to have a block representation is to expose controls for the _layout_ of that functionality. 
- There is no straightforward interface affordance for these "ghost" or "dynamic" type blocks which are conditionally applied in various views and I think Manos has captured all the difficulties around this in his comments.
- This is definitely a UI/UX design challenge needing solved. In the meantime, perhaps what we should do is ensure that any default functionality is still rendered via the compatibility layer (i.e. upsells).

**Comment by gigitux** [comment]
>  In the meantime, perhaps what we should do is ensure that any default functionality is still rendered via the compatibility layer (i.e. upsells).

Yes! We already did it, and it has been merged https://github.com/woocommerce/woocommerce-blocks/pull/9636.

**Comment by tjcafferkey** [comment]
>Manos has raised very valid points

Agreed, and regarding the conditional blocks (do we have a term for these types of blocks yet?) I think it's worth looping Gutenberg into any discussions sooner rather than later to collaborate with.

There have been times where we've come across problems which we have solved ourselves (e.g. loading block templates from a plugin) only for someone from Gutenberg to ask why we haven't created an issue for this in the repo. If we don't tell them this is an issue, how will they ever know? I can think of plenty of other scenarios outside of the context of WooCommerce where these conditional blocks may (or may not) need to exist so it can be definitely useful for the wider community and ecosystem.

**Comment by jarekmorawski** [comment]
Thanks for the ping. This is a problem that, if solved correctly, will elevate the overall store editing experience.

> I am tempted to think that it would be useful to introduce product-type templates.

I'd go even further and say that letting users edit pages for individual products would be even more helpful. In addition to Upsells and Cross-sells, plugins like Product Add-ons add custom fields to products. In theory, they're product specific. However, they're only relevant on the storefront. Why would merchants add them in the product form using regular buttons and inputs when they'd just drop a text input field into the page template?

This has not been solved in core yet, but the first step will be the new flows for template vs. content editing. In our case, I imagine a user editing the non-tangible data in the product form, then switching to the site editor (all within the same UI) to edit the page design, including custom fields and blocks. This could even mean editing the description visually right on the page rather than on a dedicated screen.

> might be a great use-case for things like the "upsell" configuration in the editor where the design of the upsell block could potentially be modified.

Do you mean that the user could tweak the upsells section's appearance from within the product form? I discussed this with the core team earlier today, and they seemed torn. There's an overarching motion that the management and editing use modes are and shall remain separate. You'd create a product in the form, then go to the editor to edit the page's design. Different tools for different tasks.

However, elements like Upsells or Add-ons live in the grey area between the two... and it's probably where we'll need to take a strong position and stick with it while working on the rest of the experience and setting the example for 3PDs.

**Comment by Brianmitchtay** [comment]
6418820-zen

Both I and the merchant were surprised to see that upsells are not shown when added to a product.

It looks like the PR here has been merged into 10.4.0, and WC core is only on 10.0.6. I'll explain to them that they can install the WooCommerce Blocks plugin to get the latest version that has the upsells reactivated.

**Comment by gigitux** [comment]
Closing this issue! Upselling products is a variant of the Product Collection
